addons/udes_stock_refactoring/models/stock_move.py

```diff
@@ -0,0 +1,355 @@
+# -*- coding: utf-8 -*-
+
+from odoo import api, models, fields, _
+from odoo.exceptions import UserError
+
+from .refactor import REFACTOR_REGISTRY
+
+import logging
+
+_logger = logging.getLogger(__name__)
+
+
+# Map move state to the refactor stage.
+U_STOCK_REFACTOR_STAGES = {
+    "confirmed": "confirm",
+    "waiting": "confirm",
+    "assigned": "assign",
+    "partially_available": "assign",
+    "done": "validate",
+}
+
+
+class StockMove(models.Model):
+    _inherit = "stock.move"
+
+    def _get_refactor_class(self, picking_type, stage):
+        """
+        Return the relevant refactor class from the provided 
+        picking type and stage, if applicable.
+        """
+        try:
+            refactor_action = False
+
+            if stage == "confirm":
+                refactor_action = picking_type.u_post_confirm_action
+            elif stage == "assign":
+                refactor_action = picking_type.u_post_assign_action
+            elif stage == "validate":
+                refactor_action = picking_type.u_post_validate_action
+
+            if refactor_action:
+                refactor_class = REFACTOR_REGISTRY[refactor_action]
+                return refactor_class(self.env)
+
+            return refactor_action
+        except KeyError:
```

Comment:
Can we make the try except as small as possible :) Or use a get and if :)

Reply:
Yeah good point. Have now moved the try/except to only cover when it tries to actually find the class. Have also added a `stage_post_action_dict` in `_get_refactor_class` to replace the if statements.